src/util/io.h

```diff
@@ -5,20 +5,42 @@
  */
 #ifndef UTIL_IO_H
 #define UTIL_IO_H
+
+#ifdef _WIN32
+#include <BaseTsd.h>
+#include <winsock2.h>
+#include <ws2tcpip.h>
+typedef SSIZE_T ssize_t;
+#define _HOST_NAME_MAX MAX_COMPUTERNAME_LENGTH
+
+#else
 #include <arpa/inet.h>
 #include <errno.h>
 #include <sys/socket.h>
 #include <sys/un.h>
+#define _HOST_NAME_MAX _POSIX_HOST_NAME_MAX
+#define SOCKET int
+#define INVALID_SOCKET - 1
+#define SOCKET_ERROR - 1
+#endif
 
 #include "tss2_tpm2_types.h"
 
-#define SOCKET int
-#define TEMP_RETRY(exp) \
-({  int __ret; \
+#ifdef _WIN32
+#define TEMP_RETRY(dest, exp) \
+{   int __ret; \
     do { \
         __ret = exp; \
-    } while (__ret == -1 && errno == EINTR); \
-    __ret; })
+    } while (__ret == SOCKET_ERROR && WSAGetLastError() == WSAEINTR); \
+    dest = __ret; }
```

Comment:
Why do we need to change this? Doesn't Visual Studio support return values from macros?

Reply:
Error codes made by Windows Sockets are not made available through errno, they have to be retrieved by using WSAGetLastError() instead. More information on this can be found here: https://docs.microsoft.com/en-us/windows/desktop/WinSock/error-codes-errno-h-errno-and-wsagetlasterror-2.

Reply:
Just realized I misunderstood what you were asking about in this comment, sorry about that. I explained this change in the commit message:
"Rewrite the TEMP_RETRY macro. Previously this macro was a statement
expression; statement expressions are part of the non-standard gcc
extension statement-exprs."

I have found that even with the Clang LLVM compiler, visual studio does not support statement expressions. More information about statement expressions can be found here: https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html